crates/uv-distribution/src/source/mod.rs

```diff
@@ -396,6 +396,8 @@ impl<'a, T: BuildContext> SourceDistributionBuilder<'a, T> {
         hashes: HashPolicy<'_>,
         client: &ManagedClient<'_>,
     ) -> Result<BuiltWheelMetadata, Error> {
+        let _lock = cache_shard.lock().map_err(Error::CacheWrite)?;
```

Comment:
Realizing... I think I need this to be async?

Reply:
Looks like you hold it across `await` points, so yeah.

Reply:
It’s not a mutex though, it’s a flock. I’d need to do some research.

Reply:
Hm.. aren't you still yielding control? So some other task could hit this lock and block without awaiting and returning control to the event loop and consequently to this task? e.g.

- Task A takes lock
- Task A yields to event loop
- Task B attempts to take lock
- Lock is not available so Task B is blocked but has not yielded to the event loop
- Task A is never returned to
- Deadlock

Reply:
Wrapped it in a tokio task.

Reply:
Yeah I agree with this.